driver/custom_endpoint_proxy.cc

```diff
@@ -0,0 +1,172 @@
+  }
+
+  this->custom_endpoint_host = host;
+  MYLOG_TRACE(this->logger, 0, "Detected a connection request to a custom endpoint URL: '%s'", host);
+
+  this->custom_endpoint_id = RDS_UTILS::get_rds_cluster_id(host);
+
+  if (this->custom_endpoint_id.empty()) {
+    this->set_custom_error_message("Unable to parse custom endpoint identifier from URL.");
+    return false;
+  }
+
+  this->region = ds->opt_CUSTOM_ENDPOINT_REGION ? static_cast<const char*>(ds->opt_CUSTOM_ENDPOINT_REGION)
+                                                : RDS_UTILS::get_rds_region(host);
+  if (this->region.empty()) {
+    this->set_custom_error_message(
+        "Unable to determine connection region. If you are using a non-standard RDS URL, please set the "
+        "'custom_endpoint_region' property");
+    return false;
+  }
+
+  const std::shared_ptr<CUSTOM_ENDPOINT_MONITOR> monitor = create_monitor_if_absent(ds);
+  if (this->should_wait_for_info) {
+    // If needed, wait a short time for custom endpoint info to be discovered.
+    this->wait_for_custom_endpoint_info(monitor);
+  }
+  return this->next_proxy->connect(host, user, password, database, port, socket, flags);
+}
+
+int CUSTOM_ENDPOINT_PROXY::query(const char* q) {
+  if (!this->custom_endpoint_host.empty()) {
+    const std::shared_ptr<CUSTOM_ENDPOINT_MONITOR> monitor = create_monitor_if_absent(ds);
+    if (this->should_wait_for_info) {
+      // If needed, wait a short time for custom endpoint info to be discovered.
+      this->wait_for_custom_endpoint_info(monitor);
+    }
+  }
+
+  return next_proxy->query(q);
+}
+
+int CUSTOM_ENDPOINT_PROXY::real_query(const char* q, unsigned long length) {
+  if (!this->custom_endpoint_host.empty()) {
+    const std::shared_ptr<CUSTOM_ENDPOINT_MONITOR> monitor = create_monitor_if_absent(ds);
+    if (this->should_wait_for_info) {
+      // If needed, wait a short time for custom endpoint info to be discovered.
+      this->wait_for_custom_endpoint_info(monitor);
+    }
+  }
+
+  return next_proxy->real_query(q, length);
+}
+
+void CUSTOM_ENDPOINT_PROXY::wait_for_custom_endpoint_info(std::shared_ptr<CUSTOM_ENDPOINT_MONITOR> monitor) {
+  bool has_custom_endpoint_info = monitor->has_custom_endpoint_info();
+
+  if (has_custom_endpoint_info) {
+    return;
+  }
+
+  // Wait for the monitor to place the custom endpoint info in the cache. This ensures other plugins get accurate
+  // custom endpoint info.
+  MYLOG_TRACE(this->logger, 0,
+              "Custom endpoint info for '%s' was not found. Waiting %dms for the endpoint monitor to fetch info...",
+              this->custom_endpoint_host.c_str(), this->wait_on_cached_info_duration_ms)
+
+  const auto wait_for_endpoint_info_timeout_nanos =
+      std::chrono::steady_clock::now() + std::chrono::duration_cast<std::chrono::nanoseconds>(
+                                             std::chrono::milliseconds(this->wait_on_cached_info_duration_ms));
```

Comment:
Is a cast to nanoseconds required? Should be the same if you just add in the milliseconds